Automattic/studio
# Studio: Add a component that handles a site import

Related to https://github.com/Automattic/dotcom-forge/issues/8160

## Proposed Changes

This PR adds import functionality to the studio app.

Keynotes:
1. The implementation has been done on the main process, as it's easier to handle files.
2. Import functionality is structured as follows :
    a. **A backup handler:** A class responsible for handling the different compressed format
    b. **A backup manager:** Responsible for running through the validators, and selecting the appropriate importer
    b. **Validators:** For now, it's only for the Jetpack Backup format, but we can add as many validators as we want. The validator is responsible for determining, based on the files, if each type of format meets its different constraints, and preparing the data for the importer ( based on files we want to filter out, etc )
    c. **Importers:** For now, it's only for the Jetpack Backup format, but we can add as many importers as we want. The importer is responsible for reading a custom format we have implemented and performing the import. It can further act upon files if needed.


## Testing Instructions



### Pre - test

1. Discrepancy script
```bash
#!/bin/bash

if [ "$#" -ne 2 ]; then
	echo "Usage: $0 <path_to_original_backup> <path_to_imported_site>"
	exit 1
fi

ORIGINAL_BACKUP="$1"
IMPORTED_SITE="$2"
LOG_FILE="import_test_$(date +%Y%m%d_%H%M%S).log"

# Function to compare directories
compare_directories() {
	local dir1="$1"
	local dir2="$2"
	local exclude_pattern="$3"

	diff_output=$(diff -r --exclude="$exclude_pattern" "$dir1" "$dir2")

	if [ -z "$diff_output" ]; then
		echo "No differences found in $dir1" | tee -a "$LOG_FILE"
	else
		echo "Differences found in $dir1:" | tee -a "$LOG_FILE"
		echo "$diff_output" | tee -a "$LOG_FILE"
	fi
}

echo "Starting import test at $(date)" | tee "$LOG_FILE"

# Compare wp-content directories
echo "Comparing wp-content directories..." | tee -a "$LOG_FILE"
compare_directories "$ORIGINAL_BACKUP/wp-content" "$IMPORTED_SITE/wp-content" "*.log"


# Check for studio.json
if [ -f "$ORIGINAL_BACKUP/studio.json" ]; then
	echo "Comparing studio.json..." | tee -a "$LOG_FILE"
	diff "$ORIGINAL_BACKUP/studio.json" "$IMPORTED_SITE/studio.json" | tee -a "$LOG_FILE"
	compare_md5 "$ORIGINAL_BACKUP/studio.json" "$IMPORTED_SITE/studio.json"
else
	echo "studio.json not found in original backup" | tee -a "$LOG_FILE"
fi

echo "Comparison completed. Results logged in $LOG_FILE" | tee -a "$LOG_FILE"
```

2. Patch testing UI:
```typescript
diff --git a/src/components/content-tab-import-export.tsx b/src/components/content-tab-import-export.tsx
index 436c8bcc..30acc637 100644
--- a/src/components/content-tab-import-export.tsx
+++ b/src/components/content-tab-import-export.tsx
@@ -1,6 +1,46 @@
+import { useEffect, useState } from 'react';
+import { getIpcApi } from '../lib/get-ipc-api';
+import { BackupArchiveInfo } from '../lib/import-export/import/types';
+
 interface ContentTabImportExportProps {
 	selectedSite: SiteDetails;
 }
-export function ContentTabImportExport( _props: ContentTabImportExportProps ) {
-	return null;
+export function ContentTabImportExport( { selectedSite }: ContentTabImportExportProps ) {
+	const [ file, setFile ] = useState< File | null >( null );
+
+	const handleFileChange = ( e: React.ChangeEvent< HTMLInputElement > ) => {
+		const selectedFile = e.target.files ? e.target.files[ 0 ] : null;
+		if ( selectedFile ) {
+			setFile( selectedFile );
+		}
+	};
+
+	useEffect( () => {
+		if ( file ) {
+			console.log( `File ready for import: ${ file.name }` );
+		}
+	}, [ file ] );
+
+	const handleSubmit = () => {
+		if ( file ) {
+			try {
+				const backupFile: BackupArchiveInfo = {
+					type: file.type,
+					path: file.path,
+				};
+				getIpcApi().importSite( { id: selectedSite.id, backupFile } );
+			} catch ( error ) {
+				console.error( 'Error importing site:', error );
+			}
+		} else {
+			console.warn( 'No file selected for import' );
+		}
+	};
+
+	return (
+		<div>
+			<input type="file" onChange={ handleFileChange } />
+			<button onClick={ handleSubmit }>Submit</button>
+		</div>
+	);
 }

```


#### Test 1: Files imported successfully
1. Apply the above diff
4. Create a new site.
5. Export a jetpack backup from wpcom.
6. Visit `wp-admin` and check your plugins.
7. Try to import the backup
8. Ensure that by visiting wp-admin, under plugins you see something like below
![2024-07-10T21:12:52,329348617+03:00](https://github.com/Automattic/studio/assets/497103/49adb1ee-b345-4fe2-a59b-9fd979ad88c9)

#### Test 2: Check for discrepancies between the files
1. Run the script above using `./compare.sh <path with downloaded zip>/extracted <path of your site>`
2. Ensure that you see something like below:
```Comparing wp-content directories...
Differences found in /home/wild/Downloads/extracted/wp-content:
Only in /home/wild/Studio/my-bold-website/wp-content: database
Only in /home/wild/Studio/my-bold-website/wp-content: db.php
diff -r '--exclude=*.log' /home/wild/Downloads/extracted/wp-content/index.php /home/wild/Studio/my-bold-website/wp-content/index.php
2c2
< //silence is golden
---
> // Silence is golden.
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-32bit-integer-warnings.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-allowed-redirect-hosts.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-check-theme-availability.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-dns-functions.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-permalinks.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-sqlite.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-thumbnails.php
Only in /home/wild/Downloads/extracted/wp-content/mu-plugins: index.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: sqlite-database-integration
Only in /home/wild/Studio/my-bold-website/wp-content/plugins: hello.php
Only in /home/wild/Studio/my-bold-website/wp-content/themes: twentytwentyfour
Only in /home/wild/Studio/my-bold-website/wp-content/themes: twentytwentythree
Only in /home/wild/Studio/my-bold-website/wp-content/themes: twentytwentytwo
Only in /home/wild/Studio/my-bold-website/wp-content: uploads
studio.json not found in original backup
Comparison completed. Results logged in import_test_20240710_213658.log
```
3. Ensure that none of the files included in the extracted backup is missing
## Pre-merge Checklist
